hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/batching/PartitionAwareCallableBatchingRunnable.java

```diff
@@ -0,0 +1,128 @@
+ * <ul>
+ * <li>Runnable assumes to be run on a partition-thread. If not run on a partition-thread a RuntimeException will be thrown.</li>
+ * <li>This optimization highly relies on the threading-model and has been implemented to speed up invocations that span all
+ * partitions, e.g. HD-index-query, or potentially many more like IMap.size(), etc.</li>
+ * </ul>
+ *
+ * @see PartitionAwareCallable
+ * @see PartitionAwareCallableFactory
+ * @since 3.9
+ */
+public class PartitionAwareCallableBatchingRunnable implements Runnable {
+
+    private final PartitionAwareCallableFactory factory;
+    private final IPartitionService partitionService;
+    private final Address memberAddress;
+    private final int partitionThreadCount;
+
+    private final CopyOnWriteArrayList results = new CopyOnWriteArrayList();
+    private final AtomicInteger finished = new AtomicInteger(0);
+
+    private final ResultFuture future;
+
+    /**
+     * @param nodeEngine nodeEngine
+     * @param factory    factory of PartitionAwareCallable to be run on the partition thread.
+     */
+    public PartitionAwareCallableBatchingRunnable(NodeEngine nodeEngine, PartitionAwareCallableFactory factory) {
+        this.factory = factory;
+        this.partitionService = nodeEngine.getPartitionService();
+        this.memberAddress = nodeEngine.getThisAddress();
+        this.partitionThreadCount = ((OperationServiceImpl) nodeEngine.getOperationService()).getPartitionThreadCount();
+        this.future = new ResultFuture(nodeEngine, nodeEngine.getLogger(getClass()));
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void run() {
+        if (!(Thread.currentThread() instanceof PartitionOperationThread)) {
+            throw new IllegalArgumentException("The PartitionAwareCallableBatchingRunnable has to be run on a partition-thread");
+        }
+        PartitionOperationThread currentThread = ((PartitionOperationThread) Thread.currentThread());
+        int partitionThreadId = currentThread.getThreadId();
+        try {
+            int currentPartitionId = partitionThreadId;
+            IPartition[] partitions = partitionService.getPartitions();
+            runSequentially(currentThread, currentPartitionId, partitions);
+        } finally {
+            int value = finished.incrementAndGet();
+            if (!future.isDone() && value == partitionThreadCount) {
+                future.setResult(results);
+            }
+        }
+    }
+
+    private void runSequentially(PartitionOperationThread currentThread, int currentPartitionId, IPartition[] partitions) {
+        while (currentPartitionId < partitions.length) {
+            final IPartition partition = partitions[currentPartitionId];
+            if (currentThread.isInterrupted()) {
+                future.cancel(true);
+                break;
+            }
+            PartitionAwareCallable task = factory.create();
+            if (memberAddress.equals(partition.getOwnerOrNull())) {
```

Comment:
`if (partition.isLocal())` seems a concise form of `if (memberAddress.equals(partition.getOwnerOrNull()))`

Reply:
Nice improvement! Fixed.